### PR TITLE
Feature/contributing

### PR DESCRIPTION
**Pull request by kevinahuber**
## Contributing!

This is a _start_ for our contributing guidelines. 

We should keep the root of docs relatively clean, with the readme directing people to the appropriate place. 

I intend on adding a `reviewing` section for guidelines of our reviewers with things like when to squash, responding guideline, etc. 

This addresses #9 

NOTE: I moved the quickstart guide (`contributing-quickstart.md` -> `contributing/quickstart.md`)
This *will* break some links we have already been distributing

### CODE REVIEW

**Comment by samsontmr** [comment]
Some sections are covered by my PR #9. Should I close mine? Or we can link to the other markdowns instead of duplicating the sections.

**Comment by kevinahuber** [comment]
Yep! Haha my redundancy of your work just proves the need for it :)

I'll pull your concepts into mine, then if you can review it afterwards to confirm that I haven't kept anything out.

**Comment by kevinahuber** [comment]
@samsontmr Take a look

File: contributing/guidelines.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+### Step 3: Do work and commit
+
+- One commit per logical change
+- Be descriptive (no `Update README.md`)
+- [Commitizen](https://github.com/commitizen/cz-cli) is a wonderful tool that helps keep commits organized. If you are frequently contributing, we highly recommend using it.
+
+### Step 4: Pull Request
+- Pull request back into master and request review. All code must meet [code standards](#code), pass integration tests, and provide a meaningful contribution.
+- Tag the issues that the pull request is fulfilling
```
Comment (samsontmr) [inline contributing/guidelines.md]: Add "using a GitHub keyword (e.g. `Closes #1`)"  here? Keyword autocloses issue upon PR merge.

File: contributing/labeling.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+* `Epic`: Big issue to be broken down into smaller tasks
+
+### Priority (`p.*`)
+* `p.Critical`: To be handled as soon as possible, usually a bug or security loophole
+* `p.High`: Issues corresponding to the upcoming `Milestone`
+* `p.Medium`: Issues corresponding to a future `Milestone`
+* `p.Low`: Good-to-have features not in the roadmap
+
+### Difficulty (`d.*`)
```
Comment (anoff) [inline contributing/labeling.md]: don't you think we could go easier on the amount/granularity of labels? imo `FirstTimer/Beginner` is enough as a label. I imagine it being difficult to differentiate Intermediate, Advanced.

File: contributing/labeling.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Label Definitions
+
+* `Epic`: Big issue to be broken down into smaller tasks
+
+### Priority (`p.*`)
```
Comment (anoff) [inline contributing/labeling.md]: `High/Low` would suffice imo. We should use the `Milestone` links which already indicates that an issue has a higher relevancy than those that do not have a milestone attached to it. `High` prio as an additional label might only be necessary for bugs.

Comment (samsontmr) [reply contributing/labeling.md]: Good point.

File: contributing/labeling.md (hunk 2 above)
Comment (samsontmr) [reply contributing/labeling.md]: Okay with collapsing intermediate and advanced but i was thinking `FirstTimer` issues would be very very simple changes, e.g. minor bugs, spelling errors, and `beginner` would be more like those in the PySyft repo at the moment. Many OSS projects have this distinction to help new contributors find easy issues to get started with.

Comment (anoff) [reply contributing/labeling.md]: Ok! Would you think that just firsttimer and beginner is enough or is there a necessity to highlight issues that are neither of those?

Comment (samsontmr) [reply contributing/labeling.md]: What do you think of leaving Advanced for issues that require more indepth research? We can leave normal issues untagged in this case.

**Comment by adrianbrink** [comment]
I would recommend that instead of developing our own convention for branching and merging we should adopt [git-flow](http://nvie.com/posts/a-successful-git-branching-model/). 

It is a very simple yet very effective branching strategy that allows a lot of people to contribute.

As an extensions I would suggest that everyone has to include their own name in the branch name. Furthermore I would suggest that we every PR has to be tagged with 'bug', 'feature', 'docs', ... .

**An example:**

I want to fix a bug a discovered on `develop`.

I checkout my branch from develop using `git checkout -b feature/adrian-bug_fix`.

I do some development, I commit it, and then push that branch to the main repository.

Now I open a PR from that branch onto `develop` and tag it with `bug`.


**Unresolved questions:**
How will this flow work from forks?

File: contributing/labeling.md (hunk 2 above)
Comment (anoff) [reply contributing/labeling.md]: imho `labels` should not be used to categorise things but allow selection based on use cases. If you have a label for everything, they end up being worthless.

**Comment by anoff** [comment]
Do we need `develop` and `master`? In my experience we often dropped one of the branches because they were only there to satisfy a process. Never had an actual use case for it.

I'm favouring [gitlab flow](https://about.gitlab.com/2014/09/29/gitlab-flow/), read the link for some strong arguments :)

**Comment by adrianbrink** [comment]
We are currently using it for Tendermint and Ethermint and we do all the work on `develop` and then merge it into master. No one ever touches `master` directly and we so far it has prevented us from broken code on master.

**Comment by anoff** [comment]
What would happen if `master` breaks? In my experience: a pipeline fails, code does net get deployed, some alerts go off and people fix it. Same thing happens if you break `develop`. If you are working with short iterations issues should get fixed (fail forward).

The question is if anyone has ever had bad experiences with not having a develop branch. Otherwise I don't see a reason to choose the complex workflow over the simple one?

File: contributing/labeling.md (hunk 2 above)
Comment (samsontmr) [reply contributing/labeling.md]: Hm good point! Let's remove Intermediate and Advanced then.

**Review by iamtrask — COMMENTED** [review]
I agree on what seems to be the final concensus on labels as well as adrian's thoughts on not reinventing the wheel. Only big thing to add here is that we want to make sure we tie it into the Contributor Quickstart Guide (meaning link from the quickstart to this doc) so that there's a 100% obvious "this is the next step" for total noobs to walk through to get up to speed with the project. The contents of this doc are essential and we wouldn't want people to miss them when onboarding (would cause headaches explaing this doc over and over when merging PRs)

**Comment by iamtrask** [comment]
@cereallarceny anything more to add prior to a merge? (also... we still have some conflicts to address)

**Comment by cereallarceny** [comment]
None for me in particular @iamtrask - I've mostly just been following along for the ride.  Personally, I've always loved git-flow.  That's my preference for workflow since it's pretty universally used in every major project I've ever done.  I think the biggest argument to use git-flow is actually just because it's so commonly known and would be most intuitive for new contributors.  There's no need to learn another Github contribution model when there's already a great one established.  However, don't let that stop us from defining our own system.

Maybe I also misunderstand the argument here.  That's fairly likely.  :)

File: contributing/guidelines.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- Issue titles should be concise yet descriptive.<br>
+  **Example** Instead of `Newbie question, please help`, use `How do I set up git to ignore test files?`
+
+- Phrasing should match the main purpose of the issue.<br>
+  For example, if it is a bug report, the issue title should sound like a bug report
+  **Example** `Option 'other' is missing from the dropdown` instead of a feature request `Add 'other' option to the dropdown`.
+
+### Step 2: Branch (and maybe fork)
+
+- Pull requests to master only.
```
Comment (anoff) [inline contributing/guidelines.md]: this should be changed to `develop` if we follow gitflow

**Comment by kevinahuber** [comment]
Up for gitflow + branch naming. Excellent feedback. So
- [x] Resolve Conflicts
- [x] Difficulty to `FirstTimer` and `Beginner` 
- [x] Priority to `High/Low`
- [x] Add gitflow language and links
- [x] Add like to guidelines at the end of quickstart

**Comment by kevinahuber** [comment]
@adrianbrink 
> As an extensions I would suggest that everyone has to include their own name in the branch name. 

I don't agree with this- it would be redundant if the user is PR'ing off of a fork, and it gets gnarly if there is more than one person working on a feature. What are your thoughts behind it?

**Comment by kevinahuber** [comment]
Olright! I think I've addressed everything. If you have things to *add*, but agree with this general framework, please file an issue for a future PR. 

Two things to do once this is merged:
- [x] Change all quickstart links in slack and elsewhere
- [x] Add develop and switch PRs base to develop

**Comment by iamtrask** [comment]
@cereallarceny love to get your review on this ticket before we merge :)

**Review by cereallarceny — APPROVED** [review]

**Pull request merged**